torch/_inductor/ir.py

```diff
@@ -4116,6 +4116,9 @@ def get_read_writes(self):
         for input in self.inputs:
             if isinstance(input, list):
                 reads.update(StarDep(x.get_name()) for x in input)
+            elif isinstance(input, ShapeAsConstantBuffer):
+                # Skip creating dependncy for symbolics as they're visible globally
+                continue
```

Comment:
I don't know what's going on here? @eellison @Chillee

Reply:
Why do we need to do this ? what was the error previously ?

Reply:
Something like ShapeAsConstant node did't implement some methods as tensorBoxes

Reply:
Would you mind printing full error ?

Reply:
Sure:
```
======================================================================
ERROR: test_cond_simple_with_int_closure_device_cuda (__main__.CondTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2983, in wrapper
    method(*args, **kwargs)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 532, in instantiated_test
    test(self, **param_kwargs)
  File "/data/users/yidi/pytorch/test/inductor/test_control_flow.py", line 242, in test_cond_simple_with_int_closure
    self._run_test(
  File "/data/users/yidi/pytorch/test/inductor/test_control_flow.py", line 216, in _run_test
    result = model(*inputs_with_predicates)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 487, in _fn
    return fn(*args, **kwargs)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 1350, in __call__
    return self._torchdynamo_orig_callable(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 1141, in __call__
    result = self._inner_convert(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 543, in __call__
    return _compile(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 963, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 694, in compile_inner
    return _compile_inner(code, one_graph, hooks, transform)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_utils_internal.py", line 87, in wrapper_function
    return function(*args, **kwargs)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 727, in _compile_inner
    out_code = transform_code_object(code, transform)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/bytecode_transformation.py", line 1337, in transform_code_object
    transformations(instructions, code_options)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 228, in _fn
    return fn(*args, **kwargs)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 656, in transform
    tracer.run()
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2795, in run
    super().run()
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 997, in run
    while self.step():
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 909, in step
    self.dispatch_table[inst.opcode](self, inst)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2986, in RETURN_VALUE
    self._return(inst)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2971, in _return
    self.output.compile_subgraph(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 1143, in compile_subgraph
    self.compile_and_call_fx_graph(tx, pass2.graph_output_vars(), root)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 1370, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 1417, in call_user_compiler
    return self._call_user_compiler(gm)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 1466, in _call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 1447, in _call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/repro/after_dynamo.py", line 130, in __call__
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/__init__.py", line 2241, in __call__
    return compile_fx(model_, inputs_, config_patches=self.config)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 1555, in compile_fx
    return aot_autograd(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/backends/common.py", line 72, in __call__
    cg = aot_module_simplified(gm, example_inputs, **self.kwargs)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1087, in aot_module_simplified
    compiled_fn = dispatch_and_compile()
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1063, in dispatch_and_compile
    compiled_fn, _ = create_aot_dispatcher_function(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 524, in create_aot_dispatcher_function
    return _create_aot_dispatcher_function(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 762, in _create_aot_dispatcher_function
    compiled_fn, fw_metadata = compiler_fn(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py", line 182, in aot_dispatch_base
    compiled_fw = compiler(fw_module, updated_flat_args)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 1381, in fw_compiler_base
    return _fw_compiler_base(model, example_inputs, is_inference)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 1452, in _fw_compiler_base
    return inner_compile(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 459, in compile_fx_inner
    return wrap_compiler_debug(_compile_fx_inner, compiler_name="inductor")(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_dynamo/repro/after_aot.py", line 85, in debug_wrapper
    inner_compiled_fn = compiler_fn(gm, example_inputs)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 643, in _compile_fx_inner
    compiled_graph = FxGraphCache.load(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/codecache.py", line 1440, in load
    compiled_graph = compile_fx_fn(
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 554, in codegen_and_compile
    compiled_graph = fx_codegen_and_compile(gm, example_inputs, **fx_kwargs)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 862, in fx_codegen_and_compile
    compiled_fn = graph.compile_to_fn()
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/graph.py", line 1965, in compile_to_fn
    return self.compile_to_module().call
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/graph.py", line 1891, in compile_to_module
    return self._compile_to_module()
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/graph.py", line 1897, in _compile_to_module
    self.codegen_with_cpp_wrapper() if self.cpp_wrapper else self.codegen()
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/graph.py", line 1832, in codegen
    self.scheduler = Scheduler(self.operations)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/scheduler.py", line 1750, in __init__
    self._init(nodes)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/scheduler.py", line 1768, in _init
    self.nodes = [self.create_scheduler_node(n) for n in nodes]
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/scheduler.py", line 1768, in <listcomp>
    self.nodes = [self.create_scheduler_node(n) for n in nodes]
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/scheduler.py", line 1887, in create_scheduler_node
    return ExternKernelSchedulerNode(self, node)
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/scheduler.py", line 824, in __init__
    self.set_read_writes(node.get_read_writes())
  File "/home/yidi/.conda/envs/pytorch/lib/python3.10/site-packages/torch/_inductor/ir.py", line 4123, in get_read_writes
    reads.add(StarDep(input.get_name()))
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
AttributeError: 'ShapeAsConstantBuffer' object has no attribute 'get_name'

```

Reply:
The naming can be also tricky to give for expressions like "u0 < 3".

Reply:
I'm not sure if this interacts poorly with https://github.com/pytorch/pytorch/pull/137744 cc @anijain2305 in case are having difficulty finding the unbacked symints. 

But doesn't look like there is an issue as we iterate over shapes.